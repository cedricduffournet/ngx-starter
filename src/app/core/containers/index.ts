export * from './toaster';
export * from './http-progress-bar/http-progress-bar.component';
