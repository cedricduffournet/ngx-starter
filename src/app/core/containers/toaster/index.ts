export * from './toaster.component';
export * from './toaster.model';
