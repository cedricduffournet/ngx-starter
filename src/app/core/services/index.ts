export * from './http-progress.interceptor';
export * from './http.service';
