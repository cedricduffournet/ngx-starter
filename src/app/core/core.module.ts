import { NgModule, Optional, SkipSelf } from '@angular/core';
import { CommonModule } from '@angular/common';
import { environment } from '@env/environment';
import {
  HttpClientModule,
  HttpClient,
  HTTP_INTERCEPTORS
} from '@angular/common/http';

import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { TranslateHttpLoader } from '@ngx-translate/http-loader';
import { ModalModule } from 'ngx-bootstrap/modal';
import { ToasterModule } from 'angular2-toaster';

import {
  ToasterComponent,
  HttpProgressBarComponent
} from '@app/core/containers';
import { HttpProgressInterceptor } from '@app/core/services/http-progress.interceptor';

const MODULES: any[] = [
  HttpClientModule,
  TranslateModule.forRoot({
    loader: {
      provide: TranslateLoader,
      useFactory: HttpLoaderFactory,
      deps: [HttpClient]
    }
  }),
  ModalModule.forRoot(),
  ToasterModule.forRoot(),
  CommonModule
];

export const COMPONENTS = [ToasterComponent, HttpProgressBarComponent];

@NgModule({
  imports: MODULES,
  declarations: COMPONENTS,
  exports: COMPONENTS,
  providers: [
    {
      provide: HTTP_INTERCEPTORS,
      useClass: HttpProgressInterceptor,
      multi: true
    }
  ]
})
export class CoreModule {
  constructor(
    @Optional()
    @SkipSelf()
    parentModule: CoreModule
  ) {
    if (parentModule) {
      throw new Error(
        'CoreModule is already loaded. Import only once in main AppModule.'
      );
    }
  }
}

export function HttpLoaderFactory(http: HttpClient) {
  return new TranslateHttpLoader(
    http,
    `${environment.prefixAsset}/assets/i18n/`,
    '.json'
  );
}
