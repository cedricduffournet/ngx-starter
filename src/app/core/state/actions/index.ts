import * as LayoutCoreActions from './layout.actions';
import * as ToasterActions from './toaster.actions';
import * as HttpProgressActions from './http-progress.actions';

export { LayoutCoreActions, ToasterActions, HttpProgressActions };
