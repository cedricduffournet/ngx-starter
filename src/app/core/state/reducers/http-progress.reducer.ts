import { createReducer, on } from '@ngrx/store';
import { HttpProgressActions } from '@app/core/state/actions';

export interface State {
  processing: boolean;
  processList: boolean[];
  url: string;
}

export const INITIAL_STATE: State = {
  processing: false,
  processList: [],
  url: ''
};

export const reducer = createReducer(
  INITIAL_STATE,
  on(HttpProgressActions.begin, (state, { url }) => ({
    processing: true,
    url,
    processList: [...state.processList, true]
  })),
  on(HttpProgressActions.complete, (state, { url }) => ({
    processing: false,
    url,
    processList: [...state.processList.slice(1, state.processList.length)]
  }))
);

export const getProcessing = (state: State) => state.processing;
export const getProcessList = (state: State) => state.processList;
