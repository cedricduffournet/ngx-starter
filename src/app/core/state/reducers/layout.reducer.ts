import { LayoutCoreActions } from '@app/core/state/actions';
import { appMenu } from '@app/shell/shared/menu.model';
import { createReducer, on } from '@ngrx/store';
import { Menu } from '@app/shell/shared/menu.model';

export interface State {
  showSidenav: boolean;
  menu: Menu[];
}

export const INITIAL_STATE: State = {
  showSidenav: false,
  menu: appMenu
};

export const reducer = createReducer(
  INITIAL_STATE,
  on(LayoutCoreActions.toggleMenu, state => ({
    ...state,
    showSidenav: !state.showSidenav
  }))
);

export const getShowSidenav = (state: State) => state.showSidenav;
export const getMenu = (state: State) => state.menu;
