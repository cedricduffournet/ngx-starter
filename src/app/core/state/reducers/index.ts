import { InjectionToken } from '@angular/core';
import {
  createFeatureSelector,
  createSelector,
  ActionReducerMap,
  Action
} from '@ngrx/store';
import * as fromRouter from '@ngrx/router-store';
import * as fromLayout from '@app/core/state/reducers/layout.reducer';
import * as fromToaster from '@app/core/state/reducers/toaster.reducer';
import * as fromHttpProgress from '@app/core/state/reducers/http-progress.reducer';
import * as fromAuth from '@app/authentication/state/reducers';
import { RouterStateUrl } from '@app/core/utils/custom-route-serializer';

export interface State {
  layout: fromLayout.State;
  toaster: fromToaster.State;
  http: fromHttpProgress.State;
  router: fromRouter.RouterReducerState<RouterStateUrl>;
}

export const reducers = new InjectionToken<ActionReducerMap<State, Action>>(
  'Root reducers token',
  {
    factory: () => ({
      layout: fromLayout.reducer,
      toaster: fromToaster.reducer,
      http: fromHttpProgress.reducer,
      router: fromRouter.routerReducer
    })
  }
);

export const getRouterState = createFeatureSelector<
  fromRouter.RouterReducerState<RouterStateUrl>
>('router');

export const getPageTitle = createSelector(
  getRouterState,
  router => {
    if (router.state.data && router.state.data.title) {
      return router.state.data.title;
    }
    return '';
  }
);

export const getPreviousLink = createSelector(
  getRouterState,
  state => {
    if (state.state.data && state.state.data.previousLink) {
      return state.state.data.previousLink;
    }
    return null;
  }
);

export const getLayoutState = createFeatureSelector<State, fromLayout.State>(
  'layout'
);

export const getShowSidenav = createSelector(
  getLayoutState,
  fromLayout.getShowSidenav
);

export const getMenu = createSelector(
  getLayoutState,
  fromLayout.getMenu
);

export const getUserMenu = createSelector(
  getMenu,
  fromAuth.getLoggedUser,
  (menu, loggedUser) => {
    if (!loggedUser) {
      return [];
    }
    return menu.filter(item =>
      item.roles.some(role => loggedUser.roles.indexOf(role) !== -1)
    );
  }
);

export const getToasterState = createFeatureSelector<State, fromToaster.State>(
  'toaster'
);

export const getToasterParams = createSelector(
  getToasterState,
  fromToaster.getParams
);

export const getHttpState = createFeatureSelector<
  State,
  fromHttpProgress.State
>('http');

export const getHttpProcessList = createSelector(
  getHttpState,
  fromHttpProgress.getProcessList
);

export const isHttpProcessing = createSelector(
  getHttpProcessList,
  processList => processList.length !== 0
);
