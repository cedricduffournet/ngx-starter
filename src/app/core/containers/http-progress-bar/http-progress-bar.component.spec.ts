import { TestBed, ComponentFixture } from '@angular/core/testing';

import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { Store, MemoizedSelector } from '@ngrx/store';
import { HttpProgressBarComponent } from '@app/core/containers';
import * as fromRoot from '@app/core/state/reducers';

describe('HttpProgressBar', () => {
  let fixture: ComponentFixture<HttpProgressBarComponent>;
  let component: HttpProgressBarComponent;
  let store: MockStore<fromRoot.State>;
  let processing: MemoizedSelector<fromRoot.State, boolean>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [HttpProgressBarComponent],
      providers: [provideMockStore()]
    });
    fixture = TestBed.createComponent(HttpProgressBarComponent);
    component = fixture.componentInstance;
    store = TestBed.get(Store);
    processing = store.overrideSelector(fromRoot.isHttpProcessing, false);
  });

  it('should be created', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should display progress bar', () => {
    processing.setResult(true);

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
