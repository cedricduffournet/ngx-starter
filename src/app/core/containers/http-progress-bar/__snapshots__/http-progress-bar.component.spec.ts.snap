// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpProgressBar should be created 1`] = `
<app-http-progress-bar
  processing$={[Function Store]}
  store={[Function MockStore]}
>
  
</app-http-progress-bar>
`;

exports[`HttpProgressBar should display progress bar 1`] = `
<app-http-progress-bar
  processing$={[Function Store]}
  store={[Function MockStore]}
>
  <div
    class="progress"
  >
    <div
      class="indeterminate"
    />
  </div>
</app-http-progress-bar>
`;
